### PR TITLE
Add deployment script and helpful lib functions

### PR DESCRIPTION
**Pull request by shubha-rajan**
### Description
Lib functions for setting the BASE_URL slot setting, setting a new container version, and health checks

### Checklist
- [ ] Created tests which fail without the change (if possible)
- [ ] Extended the README / documentation, if necessary

### CODE REVIEW

File: cloud/azure/bin/lib/azure.sh [hunk 1]
```diff
@@ -12,3 +14,106 @@ function azure::create_resource_group() {
...
 
+#######################################
+# Gets the app name for the terraform-managed app service deployment
+# Note: This function works assuming there is only one app service
+# application in the resource group
+# Arguments:
+#   1: The resource group name for the terraform deployment
+#######################################
+function azure::get_app_name() {
+  az webapp list --resource-group="${1}" --query "[].name" -o tsv
```
Comment (goldblatt) [inline cloud/azure/bin/lib/azure.sh:25]: does this need an index in the "[].name"

File: cloud/deploys/staging_azure/deploy [hunk 2]
```diff
@@ -0,0 +1,24 @@
+#! /usr/bin/env bash
```
Comment (goldblatt) [inline cloud/deploys/staging_azure/deploy:1]: i think the idea is that this lives in shared and that deploy has a .sh config file that includes anything we'd need here

File: cloud/deploys/dev_azure/deploy [hunk 3]
```diff
@@ -0,0 +1,21 @@
+#! /usr/bin/env bash
```
Comment (goldblatt) [inline cloud/deploys/dev_azure/deploy:1]: I think dev_azure needs a .sh file and we source that when deploy instead of having a specific deploy file

File: cloud/azure/bin/deploy [hunk 4]
```diff
@@ -1,7 +1,27 @@
 #! /usr/bin/env bash
 
-KEEP_ORIGINAL_PWD=true source "${PROJECT_BASE_DIR}/bin/lib.sh"
-source "${PROJECT_BASE_DIR}/cloud/shared/bin/lib.sh"
-source "${PROJECT_BASE_DIR}/cloud/azure/bin/lib.sh"
+source "cloud/shared/bin/lib/health.sh"
```
Comment (goldblatt) [inline cloud/azure/bin/deploy:3]: i think this lives in shared with the set-up script instead of just bin/

File: cloud/deploys/staging_azure/configure-slot-settings [hunk 5]
```diff
@@ -0,0 +1,28 @@
+#! /usr/bin/env bash
+
+source "cloud/azure/bin/lib.sh"
+
+set -e
+
+while getopts g:s flag
```
Comment (goldblatt) [inline cloud/deploys/staging_azure/configure-slot-settings:7]: : after the s here

File: cloud/deploys/staging_azure/configure-slot-settings [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    *) out::error "Unexpected getopts value ${flag}";;
+  esac
+done
+
+
+readonly APP_NAME="$(azure::get_app_name "${resource_group}")"
+readonly CANARY_URL="$(azure::get_canary_url "${resource_group}" "${APP_NAME}")"
+readonly PRIMARY_URL="$(azure::get_primary_url "${resource_group}" "${APP_NAME}")"
+
+azure::slot_setting "canary" "${APP_NAME}"   "${APP_NAME}" "BASE_URL" "https://${CANARY_URL}" "${resource_group}"
```
Comment (goldblatt) [inline cloud/deploys/staging_azure/configure-slot-settings:21]: is app name twice here

File: cloud/deploys/staging_azure/configure-slot-settings [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+azure::slot_setting "primary" "${APP_NAME}" "BASE_URL" "https://${staging_hostname}" "${resource_group}" 
+
+# If the environment is staging, then we should set the staging hostname of the canary instance to its app service URL
+# so that it matches the base URL and is also treated as a staging instance
+azure::slot_setting "canary"  "${APP_NAME}" "STAGING_HOSTNAME" "${CANARY_URL}" "${resource_group}"
```
Comment (goldblatt) [inline cloud/deploys/staging_azure/configure-slot-settings:26]: extra space here?

File: cloud/deploys/dev_azure/configure-slot-settings [hunk 8]
```diff
@@ -0,0 +1,27 @@
+#! /usr/bin/env bash
+
+source "cloud/azure/bin/lib.sh"
```
Comment (goldblatt) [inline cloud/deploys/dev_azure/configure-slot-settings:3]: why is this here instead of in a azure/bin

File: cloud/azure/bin/deploy (hunk 4 above)
Comment (goldblatt) [reply cloud/azure/bin/deploy:3]: this also implements just the application update vs the terraform changes, is the plan for it to include terraform changes?

**Review by goldblatt — CHANGES_REQUESTED** [review]

File: cloud/azure/bin/deploy (hunk 4 above)
Comment (shubha-rajan) [reply cloud/azure/bin/deploy:3]: so I have this script being called by the deploy scripts in dev_azure and staging_azure, since calling terraform will look different in dev and staging. tbh I'm pretty lost on how to structure this code so we should probably meet and work on it together at some point

File: cloud/azure/bin/lib/azure_log.sh [hunk 9]
```diff
@@ -107,16 +107,10 @@ function azure_log::initialize_log_file() {
...
 }
 
-#######################################
-# Writes the current signed in Azure user to stdout with whitespace replaced
-# by underscores for suitability as a token in the deploy log.
-#######################################
-function azure_log::get_current_user_id() {
-  az ad signed-in-user show --query mail | sed -E 's/ +/_/g'
-}
+
```
Comment (goldblatt) [inline cloud/azure/bin/lib/azure_log.sh:116]: pls remove the extra line

**Review by goldblatt — APPROVED** [review]

File: cloud/deploys/dev_azure/deploy [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+
+set -e
+
+readonly TF_VAR_FILE="cloud/deploys/dev_azure/.auto.tfvars"
```
[the remark this answers is not in the document]
Comment (shubha-rajan) [reply cloud/deploys/dev_azure/deploy:5]: ooh this file should have been deleted! Good catch
